### PR TITLE
Add nofullscreen to windowrules

### LINKED ISSUE #1074: The firefox window has strange behavior then closes
I'm not sure how you're going to reproduce it. I am trying to connect to a corporate VPN using SNX (SSL Network Extender) based on [this](https://github.com/ruyrybeyro/chrootvpn) project. What is the point, when I click "connect" - a new browser window opens. This window is responsible for managing the VPN connection by SNX Extender. 
At the beginning of a new window in tile mode (a screenshot below), when the VNP connection goes from the `initializing` state to the `connecting` state, it takes up the entire screen, but it is not a fullscreen mode. Then, probably when the VPN is connected, for some reason the window just closes.

image

```
[LOG] Executing WAYLAND_DISPLAY=wayland-0 DISPLAY=:0 /usr/lib/firefox/firefox
[LOG] Process Created with pid 17710
[LOG] Hyprctl: dispatcher exec : /usr/lib/firefox/firefox
...
[LOG] New XDG Surface created. (class: firefox)
[LOG] Registered signal for owner a162dbb0: a1294950 -> a162dc18 (owner: XDG Window)
[LOG] Registered signal for owner a162dbb0: a1294960 -> a162dc80 (owner: XDG Window)
[LOG] Registered signal for owner a162dbb0: a1294920 -> a162dce8 (owner: XDG Window)
[LOG] Registered signal for owner a162dbb0: a12c1540 -> a162e1c8 (owner: Toplevel)
[LOG] Registered signal for owner a162dbb0: a12c1550 -> a162e230 (owner: Toplevel)
[LOG] Registered signal for owner a162dbb0: a12c1560 -> a162e160 (owner: Toplevel)
[LOG] Searching for matching rules for firefox (title: Mozilla Firefox)
[LOG] Set keyboard focus to surface a143d080, with window name: Mozilla Firefox
[LOG] Window got assigned a surfaceTreeNode 0

... >(Pressed the button in Firefox to connect to the VPN)

[LOG] Set keyboard focus to surface a143d080, with window name: Remote Access Portal - Main — Mozilla Firefox
[LOG] New layer popup created from XDG window a162dbb0 -> Remote Access Portal - Main — Mozilla Firefox
[LOG] Registered signal for owner a13bc460: a16b1440 -> a13bc4e8 (owner: HyprPopup)
[LOG] Registered signal for owner a13bc460: a16b1470 -> a13bc550 (owner: HyprPopup)
[LOG] Registered signal for owner a13bc460: a16b1480 -> a13bc5b8 (owner: HyprPopup)
[LOG] Registered signal for owner a13bc460: a16b1460 -> a13bc480 (owner: HyprPopup)
[LOG] Registered signal for owner a13bc460: a16b12f0 -> a13bc620 (owner: HyprPopup)
[LOG] Popup: Unconstrained from lx ly: 0.000000 0.000000, pHyprPopup lx ly: 1725.000000 9.000000
[LOG] New XDG Popup mapped at 1725 9
[LOG] Registered signal for owner a13a4aa0: a16b1300 -> a13a4aa8 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13a4aa0: a16b12f0 -> a13a4b10 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13a4aa0: a16b1310 -> a13a4b78 (owner: SurfaceTreeNode)
[LOG] Added a new subsurface a16aeb00
[LOG] Registered signal for owner a1682930: a16aeb98 -> a1682948 (owner: Subsurface)
[LOG] Registered signal for owner a1682930: a16aeba8 -> a16829b0 (owner: Subsurface)
[LOG] Registered signal for owner a1682930: a16aeb88 -> a1682a18 (owner: Subsurface)
[LOG] Creating a surfaceTree Root! (pWindow: a162dbb0)
[LOG] XDG Popup got assigned a surfaceTreeNode a13a4aa0
[LOG] Subsurface a16aeb00 mapped
[LOG] Registered signal for owner a13736c0: a16aea70 -> a13736c8 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13736c0: a16aea60 -> a1373730 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13736c0: a16aea80 -> a1373798 (owner: SurfaceTreeNode)
[LOG] Creating a subsurface Node! (pWindow: a162dbb0)
[LOG] XDG Popup unmapped
[LOG] Callback a1373758 -> a1373750, SurfaceTreeNode removed.
[LOG] Callback a13737c0 -> a13737b8, SurfaceTreeNode removed.
[LOG] Callback a13736f0 -> a13736e8, SurfaceTreeNode removed.
[LOG] SurfaceTree Node removed
[LOG] Callback a1682a40 -> a1682a38, Subsurface removed.
[LOG] Callback a1682970 -> a1682968, Subsurface removed.
[LOG] Callback a16829d8 -> a16829d0, Subsurface removed.
[LOG] Callback a13a4b38 -> a13a4b30, SurfaceTreeNode removed.
[LOG] Callback a13a4ba0 -> a13a4b98, SurfaceTreeNode removed.
[LOG] Callback a13a4ad0 -> a13a4ac8, SurfaceTreeNode removed.
[LOG] SurfaceTree Node removed
[LOG] Destroyed popup XDG a13bc460
[LOG] Callback a13bc648 -> a13bc640, HyprPopup removed.
[LOG] Callback a13bc5e0 -> a13bc5d8, HyprPopup removed.
[LOG] Callback a13bc578 -> a13bc570, HyprPopup removed.
[LOG] Callback a13bc510 -> a13bc508, HyprPopup removed.
[LOG] Callback a13bc4a8 -> a13bc4a0, HyprPopup removed.

... >(A new Firefox instance has opened)

[LOG] New XDG Surface created. (class: firefox)
[LOG] Registered signal for owner a16ae690: a16acbd0 -> a16ae6f8 (owner: XDG Window)
[LOG] Registered signal for owner a16ae690: a16acbe0 -> a16ae760 (owner: XDG Window)
[LOG] Registered signal for owner a16ae690: a16acba0 -> a16ae7c8 (owner: XDG Window)
[LOG] Registered signal for owner a16ae690: a16ac8d0 -> a16aeca8 (owner: Toplevel)
[LOG] Registered signal for owner a16ae690: a16ac8e0 -> a16aed10 (owner: Toplevel)
[LOG] Registered signal for owner a16ae690: a16ac8f0 -> a16aec40 (owner: Toplevel)
[LOG] Searching for matching rules for firefox (title: Mozilla Firefox)
[LOG] Set keyboard focus to surface a16af780, with window name: Mozilla Firefox
[LOG] Window got assigned a surfaceTreeNode 0
[LOG] Registered signal for owner a16ae690: a16afa40 -> a16ae690 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b80 -> a16ae830 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a16acbc0 -> a16ae968 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b10 -> a16aeaa0 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b30 -> a16aea38 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b40 -> a16ae9d0 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b50 -> a16aeb08 (owner: XDG Window Late)
[LOG] Registered signal for owner a16ae690: a1679b20 -> a16ae900 (owner: XDG Window Late)
[LOG] Registered signal for owner a13a4aa0: a16afa50 -> a13a4aa8 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13a4aa0: a16afa40 -> a13a4b10 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13a4aa0: a16afa60 -> a13a4b78 (owner: SurfaceTreeNode)
[LOG] Added a new subsurface a16b12a0
[LOG] Registered signal for owner a1682930: a16b1338 -> a1682948 (owner: Subsurface)
[LOG] Registered signal for owner a1682930: a16b1348 -> a16829b0 (owner: Subsurface)
[LOG] Registered signal for owner a1682930: a16b1328 -> a1682a18 (owner: Subsurface)
[LOG] Creating a surfaceTree Root! (pWindow: a16ae690)
[LOG] Map request dispatched, monitor DP-1, xywh: 1725.000000 725.000000 1706.000000 706.000000
[LOG] Subsurface a16b12a0 mapped
[LOG] Registered signal for owner a13736c0: a16b1210 -> a13736c8 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13736c0: a16b1200 -> a1373730 (owner: SurfaceTreeNode)
[LOG] Registered signal for owner a13736c0: a16b1220 -> a1373798 (owner: SurfaceTreeNode)
[LOG] Creating a subsurface Node! (pWindow: a16ae690)
[LOG] Searching for matching rules for firefox (title: SSL Network Extender — Mozilla Firefox)
[LOG] Window a16ae690 set title to SSL Network Extender — Mozilla Firefox

[LOG] Window a16ae690 fullscreen to 0
[LOG] Maximize request for a16ae690
[LOG] Searching for matching rules for firefox (title: SSL Network Extender — Mozilla Firefox)
[LOG] Scanout mode ON set for a16ae690
[LOG] Window a16ae690 unmapped (class firefox)
[LOG] Unregistered late callbacks XDG
[LOG] Callback a16ae6b8 -> a16ae6b0, XDG Window Late removed.
[LOG] Callback a16ae858 -> a16ae850, XDG Window Late removed.
[LOG] Callback a16ae990 -> a16ae988, XDG Window Late removed.
[LOG] Callback a16aeac8 -> a16aeac0, XDG Window Late removed.
[LOG] Callback a16aea60 -> a16aea58, XDG Window Late removed.
[LOG] Callback a16ae9f8 -> a16ae9f0, XDG Window Late removed.
[LOG] Callback a16aeb30 -> a16aeb28, XDG Window Late removed.
[LOG] Callback a16ae928 -> a16ae920, XDG Window Late removed.
[LOG] Searching for matching rules for firefox (title: SSL Network Extender — Mozilla Firefox)
[LOG] Scanout mode OFF set for a16ae690
[LOG] Framebuffer created, status 36053
[LOG] On closed window, new focused candidate is a162dbb0
[LOG] Set keyboard focus to surface a143d080, with window name: Remote Access Portal - Main — Mozilla Firefox
[LOG] Destroying the SubSurface tree of unmapped window a16ae690
[LOG] Callback a1373758 -> a1373750, SurfaceTreeNode removed.
[LOG] Callback a13737c0 -> a13737b8, SurfaceTreeNode removed.
[LOG] Callback a13736f0 -> a13736e8, SurfaceTreeNode removed.
[LOG] SurfaceTree Node removed
[LOG] Callback a1682a40 -> a1682a38, Subsurface removed.
[LOG] Callback a1682970 -> a1682968, Subsurface removed.
[LOG] Callback a16829d8 -> a16829d0, Subsurface removed.
[LOG] Callback a13a4b38 -> a13a4b30, SurfaceTreeNode removed.
[LOG] Callback a13a4ba0 -> a13a4b98, SurfaceTreeNode removed.
[LOG] Callback a13a4ad0 -> a13a4ac8, SurfaceTreeNode removed.
[LOG] SurfaceTree Node removed
[LOG] Callback a16aecd0 -> a16aecc8, Toplevel removed.
[LOG] Callback a16aec68 -> a16aec60, Toplevel removed.
[LOG] Callback a16aed38 -> a16aed30, Toplevel removed.

[ERR] Error in getAppIDClass: basic_string: construction from null is not valid
[LOG] Window a16ae690 destroyed, queueing. (class )
[LOG] Callback a16ae720 -> a16ae718, XDG Window removed.
[LOG] Callback a16ae788 -> a16ae780, XDG Window removed.
[LOG] Callback a16ae7f0 -> a16ae7e8, XDG Window removed.
[LOG] Cleanup: destroyed a window

(A new Firefox instance is closed)
```

On Sway works without any problem. So, I also wanted to know if it is possible to force reject entire screen requests for certain applications?

### PR DESCRIPTION
**Pull request by DashieTM**
#### Describe your PR, what does it fix/add?
Provides the ability to disable fullscreen mode for applications.
Most used for applications that request this for no reason.
-> potentially fixes #1074

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I Have tested this in debug mode and with a few applications, more testing is always welcome.

#### Is it ready for merging, or does it need work?
A quick check whether or not I have left out some edge cases and if the rule is done properly, just applied the way the other rules were applied.

### CODE REVIEW

**Comment by vaxerski** [comment]
I think it would be way more useful if this only disabled the app requesting fullscreen, and not fullscreen altogether. If the user uses a keybind to toggle fullscreen, why should we not allow them to do so?

**Comment by DashieTM** [comment]
Sure, we can also do that instead.

**Comment by vaxerski** [comment]
> Are all the calls to setWindowFullscreen from the requests in windows.cpp?

Ye

File: src/events/Windows.cpp [hunk 1]
```diff
@@ -723,19 +725,20 @@ void Events::listener_fullscreenWindow(void* owner, void* data) {
 
     if (PWINDOW->isHidden())
         return;
+    if (!PWINDOW->m_bNoFullscreenRequest) {
```
Comment (vaxerski) [inline src/events/Windows.cpp:728]: why? make another guard.

File: src/events/Windows.cpp [hunk 2]
```diff
@@ -908,18 +911,19 @@ void Events::listener_requestMaximize(void* owner, void* data) {
     const auto PWINDOW = (CWindow*)owner;
 
     Debug::log(LOG, "Maximize request for %x", PWINDOW);
+    if (!PWINDOW->m_bNoFullscreenRequest) {
```
Comment (vaxerski) [inline src/events/Windows.cpp:914]: make a guard

File: src/events/Windows.cpp [hunk 3]
```diff
@@ -726,8 +728,9 @@ void Events::listener_fullscreenWindow(void* owner, void* data) {
         return;
     }
 
-    if (PWINDOW->isHidden())
+    if (PWINDOW->isHidden() || PWINDOW->m_bNoFullscreenRequest) {
         return;
+    }
```
Comment (vaxerski) [inline src/events/Windows.cpp:733]: style: no brackets on short ifs

File: src/events/Windows.cpp [hunk 4]
```diff
@@ -912,8 +915,11 @@ void Events::listener_NewXDGDeco(wl_listener* listener, void* data) {
 void Events::listener_requestMaximize(void* owner, void* data) {
     const auto PWINDOW = (CWindow*)owner;
 
-    Debug::log(LOG, "Maximize request for %x", PWINDOW);
+    if (PWINDOW->m_bNoFullscreenRequest) {
```
Comment (vaxerski) [inline src/events/Windows.cpp:918]: style: no brackets on short ifs + indent by 4

Comment (DashieTM) [reply src/events/Windows.cpp]: woops sorry

**Comment by vaxerski** [comment]
Make a wiki MR and I'll merge this. Thanks!

**Pull request merged**